File: apps/ios/LightBeam/Protocol/CRC32.swift
import Foundation

/// CRC-32 (IEEE) matching Rust crc32fast and web-encoder.
enum CRC32 {
    private static let table: [UInt32] = {
        (0..<256).map { i -> UInt32 in
            var c = UInt32(i)
            for _ in 0..<8 {
                c = (c & 1) != 0 ? (0xEDB8_8320 ^ (c >> 1)) : (c >> 1)
            }
            return c
        }
    }()

    static func compute(_ data: Data) -> UInt32 {
        var c: UInt32 = 0xFFFF_FFFF
        for byte in data {
            let idx = Int((c ^ UInt32(byte)) & 0xFF)
            c = table[idx] ^ (c >> 8)
        }
        return c ^ 0xFFFF_FFFF
    }
}

File: apps/android/app/src/main/java/com/goldmann/lightbeam/protocol/Crc32.kt
package com.goldmann.lightbeam.protocol

/** CRC-32 (IEEE) matching Rust crc32fast and web encoder. */
object Crc32 {
    private val table: IntArray = IntArray(256) { i ->
        var c = i
        repeat(8) {
            c = if (c and 1 != 0) 0xEDB88320.toInt() xor (c ushr 1) else c ushr 1
        }
        c
    }

    fun compute(data: ByteArray, offset: Int = 0, length: Int = data.size): Int {
        var c = 0xFFFFFFFF.toInt()
        val end = offset + length
        for (i in offset until end) {
            c = table[(c xor data[i].toInt()) and 0xFF] xor (c ushr 8)
        }
        return c xor 0xFFFFFFFF.toInt()
    }

    fun compute(data: ByteArray): Int = compute(data, 0, data.size)
}
